tcmur_work.c

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) 2020 Red Hat, Inc.
+ *
+ * This file is licensed to you under your choice of the GNU Lesser
+ * General Public License, version 2.1 or any later version (LGPLv2.1 or
+ * later), or the Apache License 2.0.
+ */
+#define _GNU_SOURCE
+#include <stdlib.h>
+#include <stdint.h>
+#include <errno.h>
+#include <pthread.h>
+
+#include "libtcmu.h"
+#include "libtcmu_log.h"
+#include "tcmur_device.h"
+#include "tcmur_work.h"
+
+struct tcmur_work *tcmur_create_work(void)
+{
+	struct tcmur_work *work;
+
+	work = calloc(1, sizeof(*work));
+	if (!work)
+		return NULL;
+
+	if (pthread_mutex_init(&work->lock, NULL))
+		goto free_work;
+
+	if (pthread_cond_init(&work->cond, NULL))
+		goto destroy_mutex;
+
+	work->state = TCMUR_WORK_STATE_STOPPED;
+	return work;
+
+destroy_mutex:
+	pthread_mutex_destroy(&work->lock);
+free_work:
+	free(work);
+	return NULL;
+}
+
+static void __tcmur_flush_work(struct tcmur_work *work)
+{
+	if (work->state == TCMUR_WORK_STATE_STOPPED)
+		return;
+
+	/*
+	 * The lock thread may need to do a handler reopen call and try to flush
+	 * itself. Just ignore.
+	 */
+	if (pthread_self() == work->thread)
+		return;
+
+	/*
+	 * Some handlers will crash if we do a cancel so we just wait.
+	 */
+	tcmu_dbg("waiting for %lu to complete\n", work->thread);
+	pthread_cond_wait(&work->cond, &work->lock);
```

Comment:
Commit myself:

For the work here we need one refcnt, just think if there has 2 threads queued, and when destroying and flushing the work, the flush could only make sure the first thread is done, the second may still running.